tests/test_flow/test_reloading.py

```diff
@@ -0,0 +1,277 @@
+import pytest
+
+import sys
+
+from bionic.exception import CodeVersioningError
+
+
+@pytest.fixture
+def tmp_module_path(tmp_path):
+    tmp_dir_name = str(tmp_path)
+    sys.path.insert(0, tmp_dir_name)
+
+    # Pytest rewrites modules and caches the compiled pyc file. The cache can
+    # interfere with the tests below. Disable the cache.
+    dont_write_bytecode = sys.dont_write_bytecode
+    sys.dont_write_bytecode = True
+
+    yield tmp_path
+
+    # Ensure that the temporary module is not loaded by a future test
+    sys.path.remove(tmp_dir_name)
+
+    sys.dont_write_bytecode = dont_write_bytecode
+
+
+@pytest.fixture
+def write_flow_file(tmp_module_path):
+    modules = set()
+    counter = 0
+
+    def _write_flow_file(module_name, contents):
+        nonlocal counter
+        counter += 1
+        (tmp_module_path / f"{module_name}.py").write_text(contents)
+        modules.add(module_name)
+
+    yield _write_flow_file
+
+    # Invalidate caches
+    for m in modules:
+        sys.modules.pop(m, None)
+
+
+@pytest.fixture
+def write_flow(write_flow_file):
+    def _write_flow(x):
+        write_flow_file(
+            "flow",
+            f"""
+import bionic as bn
+
+builder = bn.FlowBuilder('flow')
+builder.assign('x', {x})
+
+@builder
+def x_doubled(x):
+    return x*2
+
+flow = builder.build()
+        """,
+        )
+
+    return _write_flow
+
+
+def verify_flow_contents(flow, results_map):
+    assert {k: flow.get(k) for k in results_map.keys()} == results_map
+    assert {k: getattr(flow.get, k)() for k in results_map.keys()} == results_map
+
+
+@pytest.fixture(params=["reloading in place", "not reloading in place"])
+def reloading_in_place(request):
+    return request.param == "reloading in place"
+
+
+# If we are not reloading in place, we want to check not only the reloaded flow
+# but also the original flow to verify that the values have not changed.
+@pytest.fixture()
+def check_flow(reloading_in_place):
+    flow_to_expected = {}
+
+    def _check_flow(flow, expected: dict, reload: bool):
+        if reloading_in_place:
+            if reload:
+                flow.reload()
+            verify_flow_contents(flow, expected)
+        else:
+            nonlocal flow_to_expected
+            if reload:
+                flow_to_expected[flow.reloading()] = expected
+            else:
+                flow_to_expected[flow] = expected
+            for f, r in flow_to_expected.items():
+                verify_flow_contents(f, r)
+
+    return _check_flow
+
+
+def test_no_change(write_flow, check_flow):
+    write_flow(1)
+
+    from flow import flow
+
+    check_flow(flow, {"x": 1, "x_doubled": 2}, False)
```

Comment:
I like the thoroughness of these tests! On the other hand, `check_flow` is doing a lot of work and it's not clear at a glance what it's doing or what the arguments mean.

One way to handle this would be to have a stateful "tester" object with expressive methods, like
```
tester.expect_values(x=1, x_doubled=2)
tester.reload_flow()
tester.expect_values(x=1, x_doubled=2)
```

WDYT?

Reply:
Any thoughts on this?

Reply:
Updated to use tester method